### PR TITLE
Insert argument list with parameter names after completion.

### PR DESCRIPTION
**Pull request by scheglov**
It would be nice to propose variables from the context for arguments, but DAS does not have this feature. Hopefully yet.

### CODE REVIEW

File: Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java [hunk 1]
```diff
@@ -253,7 +253,7 @@ private void forceFileAnnotation(String file) {
     }
   }
 
-  public void addCompletions(@NotNull final String completionId, @NotNull final CompletionSuggestionProcessor processor) {
+  public void addCompletions(@NotNull final String completionId, @NotNull final Processor<CompletionSuggestion> processor) {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java]: yep, Processor is better. Consumer is even better when you never need to stop processing in the middle. By convention boolean returned by process() means "process further" whereas false means stop processing.

File: Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java [hunk 2]
```diff
@@ -93,33 +113,34 @@ private static LookupElement createLookupElement(Project project, CompletionSugg
...
-      //});
+      // insert handler
+      lookup = lookup.withInsertHandler(new InsertHandler<LookupElement>() {
+        @Override
+        public void handleInsert(InsertionContext context, LookupElement item) {
+          final TemplateManager templateManager = TemplateManager.getInstance(project);
+          final Template template = templateManager.createTemplate("", "");
+          // Fill the template with the parameter names.
+          final List<String> parameterNames = suggestion.getParameterNames();
+          if (parameterNames != null) {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java]: it seems to be more accurate to check `if (parameterNames != null)` earlier and simply not to add insertHandler if not needed

File: Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java [hunk 3]
```diff
@@ -93,33 +113,34 @@ private static LookupElement createLookupElement(Project project, CompletionSugg
...
-      //lookup = lookup.withInsertHandler(new InsertHandler<LookupElement>() {
-      //  @Override
-      //  public void handleInsert(InsertionContext context, LookupElement item) {
-      //    context.getDocument().insertString(context.getStartOffset(), "()");
-      //  }
-      //});
+      // insert handler
+      lookup = lookup.withInsertHandler(new InsertHandler<LookupElement>() {
+        @Override
+        public void handleInsert(InsertionContext context, LookupElement item) {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java]: As a further improvement take a look at AutoPopupController#autoPopupParameterInfo. It shows parameter info tooltip - very handy when we are filling parameters after selecting method from completion list.

**Comment by alexander-doroshko** [comment]
General comment.
Filling or not arguments placeholders after selecting method from completion list: this appears to be one of the differences in IntelliJ IDEA vs Eclipse style. I've found request to implement the same in Java (https://youtrack.jetbrains.com/issue/IDEABKL-5496), the feature is being discussed since 2008 but so far IntelliJ goes another way.
Speaking about Dart - I'm not sure the feature adds much value in case of optional parameters, even for Eclipse-style lovers. Usually you need to pass only some of optional params, so if template adds placeholders for all of them you have to do a lot of extra keystrokes deleting them. And even in case of standart parameters, I'm not sure this template saves any keystrokes.
In IntelliJ world we usually have Parameter Info Tooltip shown automatically after selecting method from completion list (insert handler is as simple as AutoPopupController#autoPopupParameterInfo).
May be we should hold a wider discussion before enabling the feature.

**Comment by scheglov** [comment]
The PR is updated to insert only parenthesis and requesting displaying parameters popup.

PTAL

**Comment by alexander-doroshko** [comment]
Does it handle method with zero parameters correctly? I see that in Java it inserts () but caret is placed after closing ')', not inside, and no tooltip.

**Comment by scheglov** [comment]
No, it didn't.
I've updated the PR to move put cursor after parenthesis if there are no parameters.

**Comment by alexander-doroshko** [comment]
Pushed.
Found some things to improve. Please file issues for cases that can't be fixed soon.

1. if you already have `foo();` and use completion to change it to `fooBar();` insert handler adds extra parentheses: `fooBar()();`. Instead if opening paren is already there (even if with spaces between method name and opening paren), completion should not add extra `()`, but simply move caret after existing `(`. Similar logic is already done for Java.

2. Do not insert parentheses if we are completing code after show/hide in import directive.

3. There are some other cases where we shouldn't insert parentheses for functions/methods. For example when completion argument of a function that expects Function at this place. There might be other cases, you should know better as your Dart experience is better.

**Pull request merged**